README.md
# ![Screenshot](screenshot.svg 'Perfect Freehand')

Draw perfect pressure-sensitive freehand strokes.

🔗 Try out a [demo](https://perfect-freehand-example.vercel.app/).

💰 Using this library in a commercial product? Consider [becoming a sponsor](https://github.com/sponsors/steveruizok?frequency=recurring&sponsor=steveruizok).

## Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [Support](#support)
- [Discussion](#discussion)
- [Author](#author)

## Installation

```bash
npm install perfect-freehand
```

or

```bash
yarn add perfect-freehand
```

## Usage

This package's default export is a function that:

- accepts an array of points and an (optional) options object
- returns a stroke outline as an array of points formatted as `[x, y]`

```js
import getStroke from 'perfect-freehand'
```

You may format your input points as array _or_ an object. In both cases, the value for pressure is optional (it will default to `.5`).

```js
getStroke([
  [0, 0, 0],
  [10, 5, 0.5],
  [20, 8, 0.3],
])

getStroke([
  { x: 0, y: 0, pressure: 0 },
  { x: 10, y: 5, pressure: 0.5 },
  { x: 20, y: 8, pressure: 0.3 },
])
```

### Options

The options object is optional, as are each of its properties.

| Property           | Type     | Default | Description                                           |
| ------------------ | -------- | ------- | ----------------------------------------------------- |
| `size`             | number   | 8       | The base size (diameter) of the stroke.               |
| `thinning`         | number   | .5      | The effect of pressure on the stroke's size.          |
| `smoothing`        | number   | .5      | How much to soften the stroke's edges.                |
| `streamline`       | number   | .5      | How much to streamline the stroke.                    |
| `simulatePressure` | boolean  | true    | Whether to simulate pressure based on velocity.       |
| `easing`           | function | t => t  | An easing function to apply to each point's pressure. |
| `start`            | { }      |         | Tapering options for the start of the line.           |
| `end`              | { }      |         | Tapering options for the end of the line.             |
| `last`             | boolean  | true    | Whether the stroke is complete.                       |

**Note:** When the `last` property is `true`, the line's end will be drawn at the last input point, rather than slightly behind it.

The `start` and `end` options accept an object:

| Property | Type     | Default | Description                                 |
| -------- | -------- | ------- | ------------------------------------------- |
| `cap`    | boolean  | true    | Whether to draw a cap.                      |
| `taper`  | number   | 0       | The distance to taper.                      |
| `easing` | function | t => t  | An easing function for the tapering effect. |

**Note:** The `cap` property has no effect when `taper` is more than zero.

```js
getStroke(myPoints, {
  size: 8,
  thinning: 0.5,
  smoothing: 0.5,
  streamline: 0.5,
  easing: t => t * t * t,
  simulatePressure: true,
  last: true,
  start: {
    taper: 20,
    easing: t => t * t * t,
  },
  end: {
    taper: 20,
    easing: t => t * t * t,
  },
})
```

> **Tip:** To create a stroke with a steady line, set the `thinning` option to `0`.

> **Tip:** To create a stroke that gets thinner with pressure instead of thicker, use a negative number for the `thinning` option.

### Rendering

While `getStroke` returns an array of points representing the outline of a stroke, it's up to you to decide how you will render these points.

The function below will turn the points returned by `getStroke` into SVG path data.

```js
function getSvgPathFromStroke(stroke) {
  if (!stroke.length) return ''

  const d = stroke.reduce(
    (acc, [x0, y0], i, arr) => {
      const [x1, y1] = arr[(i + 1) % arr.length]
      acc.push(x0, y0, (x0 + x1) / 2, (y0 + y1) / 2)
      return acc
    },
    ['M', ...stroke[0], 'Q']
  )

  d.push('Z')
  return d.join(' ')
}
```

To use this function, first use perfect-freehand to turn your input points into a stroke outline, then pass the result to `getSvgPathFromStroke`.

```js
import getStroke from 'perfect-freehand'

const myStroke = getStroke(myInputPoints)

const pathData = getSvgPathFromStroke(myStroke)
```

You could then pass this string either to an [SVG path](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d) element:

```jsx
<path d={pathData} />
```

Or, if you are rendering with HTML Canvas, you can pass the result to a [`Path2D` constructor](https://developer.mozilla.org/en-US/docs/Web/API/Path2D/Path2D#using_svg_paths)).

```js
const myPath = new Path2D(pathData)
ctx.fill(myPath)
```

### Flattening

To render a stroke as a "flattened" polygon, add the [`polygon-clipping`](https://github.com/mfogel/polygon-clipping) package and use the following function together with the `getSvgPathFromStroke`.

```js
import polygonClipping from 'polygon-clipping'

function getFlatSvgPathFromStroke(stroke) {
  const poly = polygonClipping.union([stroke])

  const d = []

  for (let face of poly) {
    for (let points of face) {
      d.push(getSvgPathFromStroke(points))
    }
  }

  return d.join(' ')
}
```

> **Tip:** For implementations in Typescript, see the example project included in this repository.

### Example

```jsx
import * as React from 'react'
import getStroke from 'perfect-freehand'
import { getSvgPathFromStroke } from './utils'

export default function Example() {
  const [points, setPoints] = React.useState()

  function handlePointerDown(e) {
    e.preventDefault()
    setPoints([[e.pageX, e.pageY, e.pressure]])
  }

  function handlePointerMove(e) {
    if (e.buttons === 1) {
      e.preventDefault()
      setPoints([...points, [e.pageX, e.pageY, e.pressure]])
    }
  }

  return (
    <svg
      onPointerDown={handlePointerDown}
      onPointerMove={handlePointerMove}
      style={{ touchAction: 'none' }}
    >
      {points && (
        <path
          d={getSvgPathFromStroke(
            getStroke(points, {
              size: 24,
              thinning: 0.5,
              smoothing: 0.5,
              streamline: 0.5,
            })
          )}
        />
      )}
    </svg>
  )
}
```

[![Edit perfect-freehand-example](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/perfect-freehand-example-biwyi?fontsize=14&hidenavigation=1&theme=dark)

### Advanced Usage

#### `StrokeOptions`

A TypeScript type for the options object.

```ts
import { StrokeOptions } from 'perfect-freehand'
```

For advanced usage, the library also exports smaller functions that `getStroke` uses to generate its SVG data. While you can use `getStroke`'s data to render strokes with an HTML canvas (via the Path2D element) or with SVG paths, these new functions will allow you to create paths in other rendering technologies.

#### `getStrokePoints`

```js
import { strokePoints } from 'perfect-freehand'
const strokePoints = getStrokePoints(rawInputPoints)
```

Accepts an array of points (formatted either as `[x, y, pressure]` or `{ x: number, y: number, pressure: number}`) and a streamline value. Returns a set of streamlined points as `[x, y, pressure, angle, distance, lengthAtPoint]`. The path's total length will be the length of the last point in the array.

#### `getOutlinePoints`

Accepts an array of points (formatted as `[x, y, pressure, angle, distance, length]`, i.e. the output of `getStrokePoints`) and returns an array of points (`[x, y]`) defining the outline of a pressure-sensitive stroke.

```js
import { getOutlinePoints } from 'perfect-freehand'
const outlinePoints = getOutlinePoints(strokePoints)
```

## Support

Please [open an issue](https://github.com/steveruizok/perfect-freehand/issues/new) for support.

## Discussion

Have an idea or casual question? Visit the [discussion page](https://github.com/steveruizok/perfect-freehand/discussions).

## Author

- [@steveruizok](https://twitter.com/steveruizok)


CHANGELOG.md
## 0.4.10

- Improves "popping" on large strokes
- Adds `cap` to `start` and `end` options.

## 0.4.9

- Prevents duplicates when parsing points.

## 0.4.8

- Fixes bug on start caps.
- Adds simple test.

## 0.4.6

- Reduce the effect of streamline by half.
- If using actual pressure, reduce the effect by half again.
- Lerp between previous and next left/right point using streamline.
- Adds check for hard turns to line end alignment.
- Removes line start alignment. This still needs to be improved, but the previous solution is not the right one.

## 0.4.5

- Fixes typings.
- Fixes dot product check for left vs right.

## 0.4.4

- Fixes typings.

## 0.4.3

- Improves caps, corners.
- Re-writes most comments.

## 0.4.2

- Fixes dots.

## 0.4.1

- Removes default tapering.

## 0.4.0

- Adds `last` option.
- Adds `start` and `end` option, each with `taper` and `easing`.
- Improves cap handling.

## 0.3.5

- Improves caps.

## 0.3.4

- Fixes bug in short strokes.

## 0.3.3

- Adds the `easing` property. This property accepts an [easing function](https://gist.github.com/gre/1650294) that will apply to all pressure measurements (real or simulated). Defaults to a linear easing (`t => t`).

## 0.3.2

- Superficial changes.

## 0.3.1

- Improves sharp corners that aren't sharp enough for caps, but are still sharp enough to confuse the distance-checking part of the algorithm.

## 0.3.0

This version has breaking changes.

- Removes polygon-clipping as a dependency. The problems it solved are no longer problems but a developer might still use it separately for aesthetic reasons.
- Removes `clipPath`.
- Removes options types other than `StrokeOptions`.
- Removes `getShortStrokeOutlinePoints`.
- Removes `pressure` option.
- Removes `minSize` and `maxSize` options.
- Adds `size` and `thinning` options.
- Renames `smooth` to `smoothing`.
- Improves caps.
- Improves dots and short strokes.
- You can now use `thinning` to create strokes that shink at high pressure as well as at low pressure. This is a normalized value based on the `size` option:
  - at `0` the `thinning` property will have no effect on a stroke's width.
  - at `1` a stroke will reach zero width at the lowest pressure and its full width (`size`) at the highest pressure
  - at `-1` a stroke will reach zero width at the highest pressure and its full width at the lowest pressure.
- Setting `thinning` to zero has the same effect as had setting the now removed `pressure` option to `false`.
- Improves code organization and comments.

## 0.2.5

- Improves caps for start and end.
- Improves handling of short moves.

## 0.2.4

- Improves sharp corners.

## 0.2.3

- Brings back `simulatePressure` until I have a better way of guessing.

## 0.2.2

- Slight fix to line starts.

## 0.2.1

- Fixes actual pressure sensitivity.

## 0.2.0

- Breaks up algorithm into smaller functions. Because `getPath` returns an SVG path data, you can use `getPath` only with the Path2D element (for HTML Canvas) or SVG paths. These new functions will allow you to create paths in other rendering technologies.

  - `getStrokePoints` will apply a streamline to the given point, calculate angle, and distances and lengths.
  - `getStrokeOutlinePoints` will generate an array of points defining the outline of the path.
  - `getShortStrokeOutlinePoints` will generate outline points for a very short stroke.
  - `clipPath` will generate a polygon (a series of faces) from the stroke.
  - `getPath` remains unchanged, and will still return an SVG path. It calls each of the above functions (more or less, depending on the mark and its options).

- Adds the `clip` option to flatten the path into a single polygon.
- Removes `type` option. We'll try to use pressure if available. To turn off pressure, set `pressure` to false.

## 0.1.3

- Removes hidden options, uses `maxSize` for velocity calculations.

## 0.1.2

- Fixes bug in pressure.

## 0.1.2

- Fixes bug with empty input array.

## 0.1.0

- Hey world.


example/state.ts
import { createState, createSelectorHook } from '@state-designer/react'
import { v4 as uuid } from 'uuid'
import { getPointer } from './hooks/useEvents'
import { Mark, ClipboardMessage } from './types'
import getStroke from 'perfect-freehand'
import polygonClipping from 'polygon-clipping'
import { copyToClipboard } from './utils'

function getSvgPathFromStroke(stroke: number[][]) {
  if (!stroke.length) return ''

  const d = stroke.reduce(
    (acc, [x0, y0], i, arr) => {
      const [x1, y1] = arr[(i + 1) % arr.length]
      acc.push(` ${x0},${y0} ${(x0 + x1) / 2},${(y0 + y1) / 2}`)
      return acc
    },
    ['M ', `${stroke[0][0]},${stroke[0][1]}`, ' Q']
  )

  d.push(' Z')

  return d
    .join('')
    .replaceAll(/(\s?[A-Z]?,?-?[0-9]*\.[0-9]{0,2})(([0-9]|e|-)*)/g, '$1')
}

function getFlatSvgPathFromStroke(stroke: number[][]) {
  const poly = polygonClipping.union([stroke] as any)

  const d = []

  for (let face of poly) {
    for (let points of face) {
      d.push(getSvgPathFromStroke(points))
    }
  }

  return d.join(' ')
}

function getStrokePath(
  mark: Mark,
  simulatePressure: boolean,
  options: AppOptions,
  last: boolean
) {
  const stroke = getStroke(mark.points, {
    ...options,
    easing: options.easing.evaluate,
    simulatePressure,
    start: {
      cap: options.capStart,
      taper: options.taperStart,
      easing: options.taperStartEasing.evaluate,
    },
    end: {
      cap: options.capEnd,
      taper: options.taperEnd,
      easing: options.taperEndEasing.evaluate,
    },
    last,
  })

  const path = options.clip
    ? getFlatSvgPathFromStroke(stroke)
    : getSvgPathFromStroke(stroke)

  return path
}

interface Easing {
  0: number
  1: number
  2: number
  3: number
  evaluate: (t: number) => number
}

interface AppOptions {
  size: number
  streamline: number
  clip: boolean
  easing: Easing
  thinning: number
  smoothing: number
  simulatePressure: boolean
  capStart: boolean
  taperStart: number
  taperStartEasing: Easing
  capEnd: boolean
  taperEnd: number
  taperEndEasing: Easing
  cssStroke: number
}

const defaultOptions: AppOptions = {
  size: 24,
  thinning: 0.5,
  smoothing: 0.5,
  streamline: 0.5,
  simulatePressure: true,
  clip: false,
  easing: {
    0: 0.25,
    1: 0.25,
    2: 0.75,
    3: 0.75,
    evaluate: t => t,
  },
  capStart: true,
  taperStart: 0,
  taperStartEasing: {
    0: 0.25,
    1: 0.25,
    2: 0.75,
    3: 0.75,
    evaluate: t => t,
  },
  capEnd: true,
  taperEnd: 0,
  taperEndEasing: {
    0: 0.25,
    1: 0.25,
    2: 0.75,
    3: 0.75,
    evaluate: t => t,
  },
  cssStroke: 0,
}

const defaultSettings = {
  penMode: false,
  darkMode: false,
  showTrace: false,
  showControls: false,
  recomputePaths: true,
}

const state = createState({
  data: {
    settings: { ...defaultSettings },
    alg: { ...defaultOptions },
    restore: [] as { clear?: boolean; marks: Mark[] }[],
    redos: [] as { clear?: boolean; marks: Mark[] }[],
    marks: [] as Mark[],
    currentMark: null as Mark | null,
    clipboardMessage: null as ClipboardMessage | null,
    lastPressure: 0,
  },
  states: {
    app: {
      initial: 'idle',
      states: {
        idle: {
          on: {
            RESET_OPTIONS: ['resetOptions', 'updatePaths'],
            CHANGED_OPTIONS: ['changeOptions', 'updatePaths'],
            CHANGED_SETTINGS: ['changeSettings'],
            TOGGLED_CONTROLS: 'toggleControls',
            EXITED_PEN_MODE: 'clearPenMode',
            LOADED: ['setup', 'setDarkMode'],
            CLEARED_CANVAS: ['clearMarks'],
            UNLOADED: 'cleanup',
            RESIZED: ['resize'],
            UNDO: ['undoMark'],
            REDO: ['redoMark'],
            TOGGLED_DARK_MODE: ['toggleDarkMode', 'setDarkMode'],
            CLEARED_CLIPBOARD_MESSAGE: 'clearClipboardMessage',
            COPIED_TO_CLIPBOARD: [
              {
                get: 'svgElement',
                if: 'hasResult',
                to: 'copying',
                else: 'alertCouldNotCopyToClipboard',
              },
            ],
            PRESSED_KEY_D: 'toggledTrace',
            PRESSED_KEY_E: ['clearMarks'],
            PRESSED_KEY_Z: [
              { if: ['metaPressed', 'shiftPressed'], do: 'redoMark' },
              { if: 'metaPressed', unless: 'shiftPressed', do: 'undoMark' },
            ],
          },
        },
        copying: {
          async: {
            await: 'copySvgToClipboard',
            onResolve: {
              do: 'alertCopiedToClipboard',
              to: 'idle',
            },
            onReject: {
              do: [
                () => window.alert('no api'),
                'alertCouldNotCopyToClipboard',
              ],
              to: 'idle',
            },
          },
        },
      },
    },
    pointer: {
      initial: 'up',
      states: {
        up: {
          on: {
            DOWNED_POINTER: [
              { if: 'inPenMode', unless: 'hasPressure', break: true },
              { if: 'hasPressure', unless: 'inPenMode', do: 'setPenMode' },
              { do: 'beginMark', to: 'down' },
            ],
          },
        },
        down: {
          on: {
            LIFTED_POINTER: { do: 'completeMark', to: 'up' },
            MOVED_POINTER: [
              { if: 'hasPressure', unless: 'inPenMode', do: 'setPenMode' },
              'addPointToMark',
            ],
          },
        },
      },
    },
  },
  onEnter: { do: 'setDarkMode' },
  results: {
    svgElement() {
      return document.getElementById('drawable-svg')
    },
  },
  conditions: {
    hasPressure() {
      const { p, type } = getPointer()
      return type === 'pen' || p !== 0
    },
    inPenMode(data) {
      return data.settings.penMode
    },
    hasCurrentMark(data) {
      return !!data.currentMark
    },
    hasResult(data, paylad, result) {
      return !!result
    },
    shiftPressed(data, payload) {
      return payload.keys.shift
    },
    metaPressed(data, payload) {
      return payload.keys.meta
    },
  },
  actions: {
    resetOptions(data) {
      data.alg = { ...defaultOptions }
    },
    changeOptions(data, payload) {
      data.alg = { ...data.alg, ...payload }
    },
    changeSettings(data, payload) {
      data.settings = { ...data.settings, ...payload }
    },
    toggleControls(data) {
      data.settings.showControls = !data.settings.showControls
    },
    toggledTrace(data) {
      data.settings.showTrace = !data.settings.showTrace
    },
    setPenMode(data) {
      data.settings.penMode = true
    },
    clearPenMode(data) {
      data.settings.penMode = false
    },
    setup(
      data,
      payload: {
        marks: Mark[]
        alg: typeof defaultOptions
        settings: typeof defaultSettings
      }
    ) {
      const { marks, alg, settings } = payload

      data.alg = { ...data.alg, ...alg }

      data.marks = marks.map(mark => ({
        ...mark,
        path: getStrokePath(mark, mark.simulatePressure, alg, true),
      }))

      data.settings = {
        ...data.settings,
        ...settings,
        penMode: false,
      }
    },
    cleanup(data) {},
    resize(data) {},
    beginMark(data) {
      const { alg } = data
      const { x, y, p, type } = getPointer()

      data.redos = []

      data.lastPressure = p

      const point = {
        x,
        y,
        pressure: p,
      }

      data.currentMark = {
        id: uuid(),
        simulatePressure: true,
        points: [point],
        path: '',
      }

      data.currentMark.path = getStrokePath(data.currentMark, true, alg, false)
    },
    addPointToMark(data) {
      const { x, y, p } = getPointer()
      const { currentMark, alg, settings } = data

      data.lastPressure = p

      currentMark!.simulatePressure = !settings.penMode

      currentMark!.points.push({
        x,
        y,
        pressure: p,
      })

      currentMark!.path = getStrokePath(
        currentMark!,
        currentMark!.simulatePressure,
        alg,
        false
      )
    },
    completeMark(data) {
      const { currentMark, alg } = data

      if (!currentMark) return

      data.marks.push({
        ...currentMark,
        path: getStrokePath(
          currentMark!,
          currentMark!.simulatePressure,
          alg,
          true
        ),
      })

      data.currentMark = null
    },
    clearMarks(data) {
      data.marks = []
      data.redos = []
    },
    loadData(data, payload: { marks: Mark[] }) {
      const { alg } = data
      data.marks = payload.marks.map(mark => ({
        ...mark,
        id: uuid(),
        path: getStrokePath(mark, mark.simulatePressure, alg, true),
      }))
    },
    undoMark(data) {
      if (data.marks.length === 0) {
        const restored = data.restore.pop()
        if (restored) data.marks = restored.marks
        return
      }

      const undid = data.marks.pop()
      if (undid) {
        data.redos.push({ marks: [undid] })
      }
    },
    redoMark(data) {
      const undid = data.redos.pop()
      if (undid) {
        data.marks.push(...undid.marks)
      }
    },
    toggleDarkMode(data) {
      data.settings.darkMode = !data.settings.darkMode
    },
    setDarkMode(data) {
      if (typeof document === 'undefined') return

      if (data.settings.darkMode) {
        document.body.classList.add('dark')
      } else {
        document.body.classList.remove('dark')
      }
    },
    updatePaths(data) {
      const { currentMark, alg, marks } = data
      for (let mark of marks) {
        mark.path = getStrokePath(mark, mark.simulatePressure, alg, true)
      }

      if (currentMark) {
        currentMark.path = getStrokePath(
          currentMark,
          currentMark.simulatePressure,
          alg,
          false
        )
      }
    },
    // Clipboard message
    alertCopiedToClipboard(data) {
      data.clipboardMessage = {
        error: false,
        message: `Copied SVG`,
      }
    },
    alertCouldNotCopyToClipboard(data) {
      data.clipboardMessage = {
        error: false,
        message: `Unable to copy SVG.`,
      }
    },
    clearClipboardMessage(data) {
      data.clipboardMessage = null
    },
  },
  asyncs: {
    async copySvgToClipboard(data, payload, result: SVGSVGElement) {
      const element = result
      const padding = 16

      // Get the SVG's bounding box
      const bbox = element.getBBox()
      const tViewBox = element.getAttribute('viewBox')
      const viewBox = [
        bbox.x - padding,
        bbox.y - padding,
        bbox.width + padding * 2,
        bbox.height + padding * 2,
      ].join(' ')

      // Save the original size
      const tW = element.getAttribute('width')
      const tH = element.getAttribute('height')

      // Resize the element to the bounding box
      element.setAttribute('viewBox', viewBox)
      element.setAttribute('width', String(bbox.width))
      element.setAttribute('height', String(bbox.height))

      // Take a snapshot of the element
      const s = new XMLSerializer()
      const svgString = s.serializeToString(element)

      // Reset the element to its original viewBox / size
      element.setAttribute('viewBox', tViewBox)
      element.setAttribute('width', tW)
      element.setAttribute('height', tH)

      // Copy to clipboard!
      try {
        navigator.clipboard.writeText(svgString)
      } catch (e) {
        copyToClipboard(svgString)
      }
    },
  },
})

export const useSelector = createSelectorHook(state)

export default state

// state.onUpdate(d => console.log(d.log[0]))
